README
NAME
    Tatsumaki - Non-blocking web framework based on Plack and AnyEvent

SYNOPSIS
      ### app.psgi
      use Tatsumaki::Error;
      use Tatsumaki::Application;
      use Tatsumaki::HTTPClient;
      use Tatsumaki::Server;

      package MainHandler;
      use parent qw(Tatsumaki::Handler);

      sub get {
          my $self = shift;
          $self->write("Hello World");
      }

      package FeedHandler;
      use parent qw(Tatsumaki::Handler);
      __PACKAGE__->asynchronous(1);

      use JSON;

      sub get {
          my($self, $query) = @_;
          my $client = Tatsumaki::HTTPClient->new;
          $client->get("http://friendfeed-api.com/v2/feed/$query", $self->async_cb(sub { $self->on_response(@_) }));
      }

      sub on_response {
          my($self, $res) = @_;
          if ($res->is_error) {
              Tatsumaki::Error::HTTP->throw(500);
          }
          my $json = JSON::decode_json($res->content);
          $self->write("Fetched " . scalar(@{$json->{entries}}) . " entries from API");
          $self->finish;
      }

      package StreamWriter;
      use parent qw(Tatsumaki::Handler);
      __PACKAGE__->asynchronous(1);

      use AnyEvent;

      sub get {
          my $self = shift;
          $self->response->content_type('text/plain');

          my $try = 0;
          my $t; $t = AE::timer 0, 0.1, sub {
              $self->stream_write("Current UNIX time is " . time . "\n");
              if ($try++ >= 10) {
                  undef $t;
                  $self->finish;
              }
          };
      }

      package main;

      my $app = Tatsumaki::Application->new([
          '/stream' => 'StreamWriter',
          '/feed/(\w+)' => 'FeedHandler',
          '/' => 'MainHandler',
      ]);
      return $app->psgi_app;

    And now run it with:

      plackup -s Twiggy app.psgi

WARNINGS
    This is considered as alpha quality software. Most of the stuff are
    undocumented since it's considered unstable and will likely to change.
    You should sometimes look at the source code or example apps in *eg*
    directory to see how this thing works.

    Feel free to hack on it and ask me if you have questions or suggestions
    at IRC: #plack on irc.perl.org.

DESCRIPTION
    Tatsumaki is a toy port of Tornado for Perl using Plack (with
    non-blocking extensions) and AnyEvent.

    It allows you to write a web application that does a immediate response
    with template rendering, IO-bound delayed response (like fetching third
    party API or XML feeds), server push streaming and long-poll Comet in a
    clean unified API.

PSGI COMPATIBILITY
    When "asynchronous" is declared in your application, you need a PSGI
    server backend that supports "psgi.streaming" response style. If your
    application does server push with "stream_write", you need a server that
    supports "psgi.nonblocking" (and "psgi.streaming") as well.

    Currently Tatsumaki asynchronous application is supposed to run on
    Twiggy, Feersum, Corona and POE::Component::Server::PSGI.

    If "asynchronous" is not used, your application is supposed to run in
    any PSGI standard environments, including blocking multiprocess
    environments like Starman or Starlet.

TATSUMAKI?
    Tatsumaki is a Japanese for Tornado. Also, it might sound familiar from
    "Tatsumaki Senpuukyaku" of Ryu from Street Fighter II if you loved the
    Capcom videogame back in the day :)

AUTHOR
    Tatsuhiko Miyagawa <miyagawa@bulknews.net>

LICENSE
    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

SEE ALSO
    AnyEvent Plack PSGI



Changes
Revision history for Perl extension Tatsumaki

        - use Moose instead. Any::Moose is deprecated. (gugod)
        - Allow more than 9 captures in path regexes (plu)

0.1013  Thu Jul  7 11:49:18 PDT 2011
        - Added $self->binary(1) in Handler classes to indicate binary responses (gbarr, mateu)
        - Allow $self->json to be set to customize the way JSON is encoded (ask)

0.1012  Sun Feb 13 20:01:46 PST 2011
        - No code changes. Updated terribly outdated Tatsumaki.pm POD

0.1011  Tue Dec 21 11:49:49 PST 2010
        - Fixed to use $r->parameters to make it work with newer Plack::Request
        - Fixed MessageQueue leak when long-poll client immediately reconnects after disconnection (mash)

0.1010  Wed Apr  7 15:48:43 PDT 2010
        - INCOMPATIBLE: '/' in the mapping matches only with '/'. If you need the catch-all behavior,
          use '/.*' instead.
        - special case favicon.ico to serve from static/ directory
        - Handlers now handle HEAD the same as GET by default

0.1009  Tue Mar  2 14:31:15 PST 2010
        - Fixed a rare condition where response starter returns something and garbage
          collection breaks.
        - Fixed dependency and depend on Twiggy not Plack::Server::AnyEvent (zamolxes)
        - Removed BlockingFallback middleware

0.1008  Sun Jan 31 14:16:02 PST 2010
        - Fixed a MessageQueue bug where long poll cleanup timer is discarded
          (Thanks to yoshimax, hiratara and jshirley)

0.1007  Tue Dec 22 10:07:18 PST 2009
        - Make Test::Warn test optional

0.1006  Fri Dec 18 06:45:57 PST 2009
        - Fixed a bug in ->psgi_app where it could potentially let $app object disappear

0.1005  Sat Dec 12 00:57:20 PST 2009
        - Fixed services conflicts with the latest Moose (hiratara)
        - Use path_info rather than path to dispatch
        - Push buffered events immediately in long-poll (hiratara)
        - Fixed params bug where multiple values are flattened (ingy)

0.1004  Tue Dec  8 07:21:24 PST 2009
        - Fixed a memory leak in MessageQueue (jshirly)
        - Added experimental methods to set and get services from Application

0.1003  Mon Nov 16 21:38:04 PST 2009
        - Revert to use Text::MicroTemplate::File for now

0.1002  Mon Nov 16 18:03:50 PST 2009
        - Remove client ID from MXHR example because it's not needed
        - Specify lowest Mouse deps to make it work
        - Delay load Text::MicroTemplate so you can use other template engines like TT2 (ingydotnet)
        - Use Text::MicroTemplate::Extended by default (faultier, typester)

0.1001  Wed Nov 11 23:30:24 PST 2009
        - use Any::Moose: now all deps should run in pure perl, with optional XS speedups

0.1000  Wed Nov 11 01:35:05 PST 2009
        - Initial version to CPAN


lib/Tatsumaki.pm
package Tatsumaki;

use strict;
use warnings;
use 5.008_001;
our $VERSION = '0.1013';

1;
__END__

=encoding utf-8

=for stopwords

=head1 NAME

Tatsumaki - Non-blocking web framework based on Plack and AnyEvent

=head1 SYNOPSIS

  ### app.psgi
  use Tatsumaki::Error;
  use Tatsumaki::Application;
  use Tatsumaki::HTTPClient;
  use Tatsumaki::Server;

  package MainHandler;
  use parent qw(Tatsumaki::Handler);

  sub get {
      my $self = shift;
      $self->write("Hello World");
  }

  package FeedHandler;
  use parent qw(Tatsumaki::Handler);
  __PACKAGE__->asynchronous(1);

  use JSON;

  sub get {
      my($self, $query) = @_;
      my $client = Tatsumaki::HTTPClient->new;
      $client->get("http://friendfeed-api.com/v2/feed/$query", $self->async_cb(sub { $self->on_response(@_) }));
  }

  sub on_response {
      my($self, $res) = @_;
      if ($res->is_error) {
          Tatsumaki::Error::HTTP->throw(500);
      }
      my $json = JSON::decode_json($res->content);
      $self->write("Fetched " . scalar(@{$json->{entries}}) . " entries from API");
      $self->finish;
  }

  package StreamWriter;
  use parent qw(Tatsumaki::Handler);
  __PACKAGE__->asynchronous(1);

  use AnyEvent;

  sub get {
      my $self = shift;
      $self->response->content_type('text/plain');

      my $try = 0;
      my $t; $t = AE::timer 0, 0.1, sub {
          $self->stream_write("Current UNIX time is " . time . "\n");
          if ($try++ >= 10) {
              undef $t;
              $self->finish;
          }
      };
  }

  package main;

  my $app = Tatsumaki::Application->new([
      '/stream' => 'StreamWriter',
      '/feed/(\w+)' => 'FeedHandler',
      '/' => 'MainHandler',
  ]);
  return $app->psgi_app;

And now run it with:

  plackup -s Twiggy app.psgi

=head1 WARNINGS

This is considered as alpha quality software. Most of the stuff are
undocumented since it's considered unstable and will likely to
change. You should sometimes look at the source code or example apps
in I<eg> directory to see how this thing works.

Feel free to hack on it and ask me if you have questions or
suggestions at IRC: #plack on irc.perl.org.

=head1 DESCRIPTION

Tatsumaki is a toy port of Tornado for Perl using Plack (with
non-blocking extensions) and AnyEvent.

It allows you to write a web application that does a immediate
response with template rendering, IO-bound delayed response (like
fetching third party API or XML feeds), server push streaming and
long-poll Comet in a clean unified API.

=head1 PSGI COMPATIBILITY

When C<asynchronous> is declared in your application, you need a PSGI
server backend that supports C<psgi.streaming> response style. If your
application does server push with C<stream_write>, you need a server
that supports C<psgi.nonblocking> (and C<psgi.streaming>) as well.

Currently Tatsumaki asynchronous application is supposed to run on
L<Twiggy>, L<Feersum>, L<Corona> and L<POE::Component::Server::PSGI>.

If C<asynchronous> is not used, your application is supposed to run in
any PSGI standard environments, including blocking multiprocess
environments like L<Starman> or L<Starlet>.

=head1 TATSUMAKI?

Tatsumaki is a Japanese for Tornado. Also, it might sound familiar
from "Tatsumaki Senpuukyaku" of Ryu from Street Fighter II if you
loved the Capcom videogame back in the day :)

=head1 AUTHOR

Tatsuhiko Miyagawa E<lt>miyagawa@bulknews.netE<gt>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

L<AnyEvent> L<Plack> L<PSGI>

=cut
